File: OlderCodes/JavaCodes/Slit.java
/*
************************************************************************
*  Slit.java: Solves the time-dependent Schroedinger equation for a    *
*   two-dimensional Gaussian wavepacket entering a slit                *
*                                                                      *
*  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
   by RH Landau, MJ Paez, and CC Bordeianu 
   Copyright Princeton University Press, Princeton, 2007.
   Electronic Materials copyright: R Landau, Oregon State Univ, 2007;
   MJ Paez, Univ Antioquia, 2007; & CC Bordeianu, Univ Bucharest, 2007
   Support by National Science Foundation                              
*                                                                      *
*  comment: Output data is saved in 3D grid format used by gnuplot.    *
*                                                                      *
************************************************************************
 */
import java.io.*;
public class Slit{


public static void main(String[] argv) throws IOException,
FileNotFoundException
{
    PrintWriter w = 
        new PrintWriter(new FileOutputStream("slit.dat"), true);
double psr[][][]=new double[91][91][2];
double psi[][][]=new double[91][91][2];         
double v[][]=new double[91][91];        
double p2[][]=new double[91][91];       
double a1,a2,dt,dx,k0x,k0y,x0,y0,x,y;
int i,j,max,n,time; 
           
// Enter a positive integer from 1(initial time)
//to 800 to get wave packet position at that time
time=800;     
//initializes the constant values and the wave packet      
        
dx   = 0.2;
dt  = 0.0025/(dx*dx);
//initial momentum, position        
        k0x  = 0.0;
        k0y  = 2.5;     
        x0   = 0.0;
        y0   = -7.0;
        max  = 90;
//initial wave function
        y   = -9.0;
    for( j=0;j<max;j++){
    
          x=-9.0;
          for( i=0;i<max;i++){
 /*         
 compute real part and imaginary part of the initial wave
 function using Euler's formula: 
 exp(i*(k0x*x+k0y*y))=cos(k0x*x+k0y*y)+i*sin((k0x*x+k0y*y)             
*/
a1 = Math.exp(-0.5*(Math.pow(x-x0,2.0)+Math.pow(y-y0,2.0)));
//real part of the initial wave function
            psr[i][j][0] = a1*Math.cos(k0x*x+k0y*y);
//imaginay part of the initial wave function
            psi[i][j][0] = a1*Math.sin(k0x*x+k0y*y);
            x          = x + dx;
                              }
          y = y + dx;
                         }   
//set up the potential slit width: 50-40=10 units      

              
          
 for( j=0;j<max;j++){           
           for( i=0;i<max;i++){ 
  if((j==33)&&((i<=39)||(i>=50)))
        v[i][j] = 0.5;       
            else  
              v[i][j] = 0.0;
                                         
                               } 
                     }

//propagate solution through time     
for(n=0;n<time;n++){
      
//compute real part of wave packet and probability              
          
for(j=1;j<max-1;j++){ 
            
    for(i=1;i<max-1;i++){     
     a2 = v[i][j]*psi[i][j][0]+2.0*dt*psi[i][j][0];
     a1 = psi[i+1][j][0]+psi[i-1][j][0]+psi[i][j+1][0]+psi[i][j-1][0];
     psr[i][j][1] = psr[i][j][0]-dt*a1+2.0*a2;
     p2[i][j] = psr[i][j][0]*psr[i][j][0]+psi[i][j][0]*psi[i][j][1];  
                          
                         }  
//at x edges derivative is zero  
       psr[0][j][1] = psr[1][j][1]; 
       psr[max][j][1] = psr[max-1][j][1];                  
                       }   
//imaginary part of wave packet is next                  
          
 for(j=1;j<max-1;j++){ 
    for(i=1;i<max-1;i++){            
            
     a2 = v[i][j]*psr[i][j][1]+2.0*dt*psr[i][j][1];
     a1 = psr[i+1][j][1]+psr[i-1][j][1]+psr[i][j-1][1]+psr[i][j+1][1];
     psi[i][j][1] = psi[i][j][0]+dt*a1-2.0*a2;   
                         }  
//at x edges derivative is zero  
     psi[0][j][1]    = psi[1][j][1]; 
     psi[max][j][1] = psi[max-1][j][1];                  
                      }         
//new iterations are now the old ones, recycle
          
for(j=1;j<max-1;j++){ 
    for(i=1;i<max-1;i++){                
     psi[i][j][0] = psi[i][j][1];
     psr[i][j][0] = psr[i][j][1];
                        }   
                     }                  
              }   

//write the probabilities plus potential
//in disk file, to give a volume sensation of the potential 

 for(j=0;j<max;j=j+3){ 
      for(i=0;i<max;i=i+2){            
            w.println(""+(p2[i][j]+v[i][j])+""); 
                          }
   
            w.println("");        
                        }
        
        
        System.out.println("data saved in slit.dat");   
         
}
}

File: OlderCodes/Fortran77codes/slit.f
c  From: "A SURVEY OF COMPUTATIONAL PHYSICS" 
c  by RH Landau, MJ Paez, and CC BORDEIANU 
c  Copyright Princeton University Press, Princeton, 2008.
c  Electronic Materials copyright: R Landau, Oregon State Univ, 2008;
c  MJ Paez, Univ Antioquia, 2008; and CC BORDEIANU, Univ Bucharest, 2008.
c  Support by National Science Foundation   
c    		
c  slit.f: Solves the time-dependent Schroedinger equation for  2-D 
c	        Gaussian wavepacket entering a slit (which takes some time)
c         Output saved in gnuplot 3D grid format
c
      Program slit
			
      	Implicit None
      	Real*8 psr(91, 91, 2), psi(91, 91, 2), v(91, 91), p2(91, 91)
      	Real*8 a1, a2, dt, dx, k0x, k0y, x0, y0, x, y
      	Integer i, j, max, n, time 
      	complex exc, zi  
				
c       input pos int proportional to t when want to view wave packet
      	Write(*, *)'Enter a positive integer from 1 (initial time)'
      	Write(*, *)'to 800 to get wave packet position at that time'
      	Read(*, *)time
      	Write(*, *)'processing data for time', time
      	Open(9, file = 'slit.dat', status = 'Unknown')
c                                initialize constants and wave packet
      	zi = cmplx(0.d0, 1.d0)
      	dx = 0.2d0
      	dt = 0.0025/(dx*dx)
c                                          initial momentum, position
      	k0x = 0.d0
      	k0y = 2.5d0      
      	x0 = 0.d0
      	y0 = -7.d0
      	max = 90
c                                               initial wave function
      	y = -9.d0
	      Do 90 j = 1, max+1
          x = -9.d0
          Do 10 i = 1, max+1
            exc = exp(zi*(k0x*x+k0y*y))
            a1 = exp(-0.5*(((x-x0))**2.+((y-y0))**2.))
c                           real, imag parts of initial wave function
            psr(i, j, 1) = real(a1*exc)
            psi(i, j, 1) = imag(a1*exc)
            x = x + dx
 10       Continue
          y = y + dx
 90   	Continue    
c                         set potential slit width = 50-40 = 10 units
c
	      Do 220 j = 1, max+1           
          Do 190 i = 1, max+1
          If ((j .eq. 35) .and. ((i .le. 40) .or. (i .ge. 51))) Then 
              v(i, j) = 0.5
            Else  
              v(i, j) = 0.
            EndIf         
 190      Continue 
 220   	Continue
c                                     propagate solution through time
      	Do 40 n = 1, time   
          Do 150 j = 2, max
            Do 50 i = 2, max
c                                                     Re psi and prob
              a2 = v(i, j)*psi(i, j, 1)+2.d0*dt*psi(i, j, 1)
              a1 = psi(i+1,j,1) + psi(i-1,j,1) + psi(i,j+1,1) 
							                                        + psi(i,j-1,1)
              psr(i, j, 2) = psr(i, j, 1) - dt*a1 + 2.*a2
                If ( n .eq. time) Then
                  p2(i,j)=psr(i,j,1)*psr(i,j,1)+psi(i,j,1)*psi(i,j,1)
                EndIf         
 50         Continue  
c                                        derivative = zero  at x edge
            psr(1, j, 2) = psr(2, j, 2) 
            psr(max+1, j, 2) = psr(max, j, 2)                  
 150      Continue    
c                                                            imag psi
          Do 160 j = 2, max
            Do 60 i = 2, max
              a2 = v(i, j)*psr(i, j, 2)+2.*dt*psr(i, j, 2)
              a1 = psr(i+1,j,2) + psr(i-1,j,2) + psr(i,j-1,2) 
							                                        + psr(i,j+1,2)
              psi(i, j, 2) = psi(i, j, 1)+dt*a1-2.*a2   
 60         Continue  
c                                        derivative = zero  at x edge
            psi(1, j, 2) = psi(2, j, 2) 
            psi(max+1, j, 2) = psi(max, j, 2)                  
 160      Continue         
c                                                     new -> old ones
          Do 180 j = 1, max+1  
            Do 70 i = 1, max+1
              psi(i, j, 1) = psi(i, j, 2)
              psr(i, j, 1) = psr(i, j, 2)
 70         Continue   
 180      Continue                  
 40   	Continue    
c                          output probabilities plus scaled potential
      	Do 200 j = 2, max, 3
          Do 210 i = 2, max, 2
            Write(9, 11)p2(i, j)+v(i, j)
 210      Continue  
          Write(9, 11)      
 200  	Continue
  11	  Format (E12.6)       
       	Close(9)
       	Stop 'data saved in slit.dat'
      End  
